library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Block.hs

```diff
@@ -0,0 +1,814 @@
+
+local :: Type -> Name -> Operand
+local = LocalReference
+
+instr :: RetInstruction m => Type -> Instruction -> m Operand
+instr typ ins = do
+  n <- fresh
+  let ref = UnName n
+  blk <- current
+  let i = stack blk
+  modifyBlock (blk {stack = (ref := ins) : i})
+  pure (local typ ref)
+
+unnminstr ::
+  ( HasState "blocks" (Map.T Name BlockState) m,
+    HasState "currentBlock" Name m,
+    HasThrow "err" Errors m
+  ) =>
+  Instruction ->
+  m ()
+unnminstr ins = do
+  blk <- current
+  let i = stack blk
+  modifyBlock (blk {stack = (Do ins) : i})
+
+terminator ::
+  ( HasState "blocks" (Map.T Name BlockState) m,
+    HasState "currentBlock" Name m,
+    HasThrow "err" Errors m
+  ) =>
+  Named Terminator ->
+  m (Named Terminator)
+terminator trm = do
+  blk <- current
+  modifyBlock (blk {term = Just trm})
+  return trm
+
+--------------------------------------------------------------------------------
+-- External linking
+--------------------------------------------------------------------------------
+
+external :: (HasState "moduleDefinitions" [Definition] m) => Type -> String -> [(Type, Name)] -> m Operand
+external retty label argtys = do
+  addDefn $
+    GlobalDefinition $
+      functionDefaults
+        { Global.parameters = ((\(ty, nm) -> Parameter ty nm []) <$> argtys, False),
+          Global.callingConvention = CC.Fast, -- TODO: Do we always want this?
+          Global.returnType = retty,
+          Global.basicBlocks = [],
+          Global.name = mkName label,
+          Global.linkage = L.External
+        }
+  return $
+    ConstantOperand $
+      C.GlobalReference
+        (Types.pointerOf (FunctionType retty (fst <$> argtys) False))
+        (mkName label)
+
+externalVar :: (HasState "moduleDefinitions" [Definition] m) => Type -> String -> [(Type, Name)] -> m Operand
+externalVar retty label argtys = do
+  addDefn $
+    GlobalDefinition $
+      functionDefaults
+        { Global.parameters =
+            ( ( \(ty, nm) ->
+                  Parameter
+                    ty
+                    nm
+                    [ParameterAttribute.NoAlias, ParameterAttribute.NoCapture]
+              )
+                <$> argtys,
+              True
+            ),
+          Global.callingConvention = CC.C, -- TODO: Do we always want this?
+          Global.returnType = retty,
+          Global.basicBlocks = [],
+          Global.name = mkName label,
+          Global.linkage = L.External
+        }
+  return $
+    ConstantOperand $
+      C.GlobalReference
+        (Types.pointerOf (FunctionType retty (fst <$> argtys) True))
+        (mkName label)
+
+--------------------------------------------------------------------------------
+-- Printing facility
+--------------------------------------------------------------------------------
+
+definePrintf :: External m => m ()
+definePrintf = do
+  let name = "printf"
+  op <- externalVar Type.i32 name [(Types.pointerOf Type.i8, "n")]
```

Comment:
For functions like `prtinf` we can use it for debugging code. I think in the long run we move away from C style `printf`, however it can be quite useful when we actaully have contracts that do any amount of computation.

As for `malloc`, we can view it as an allocation api that we can build ourselves, so should be viewed as an opaque interface to a `malloc` function

Reply:
the `Type.i8` should always be fine here, as for why the types are hte way they are, they are even that way in the llvm language reference manual for an example

https://llvm.org/docs/LangRef.html#parameter-attributes